File: lib/services/task_outcome.dart
/// Pure functions that roll a task's final outcome based on agent skills.
///
/// Agents stop being defined by levels alone — the outcome roll is where
/// Precision/Creativity/Reliability skills express themselves as real
/// gameplay consequences (bug bounce, crit gold, incomplete reset).
library;

import 'dart:math';

enum TaskOutcome {
  /// Task moves `testing → done`. Normal reward.
  clean,

  /// Task moves `testing → done` AND awards a gold bonus.
  /// Rolled only on divergent task types (`taskType` in
  /// `{architecture, product-spec, ui-design}`).
  crit,

  /// Task bounces `testing → in-progress` (bug was found). Half XP.
  bug,

  /// Task is reset back to `backlog`. Half XP; agent may pick it up again.
  incomplete;

  /// Wire-key sent over the network and persisted on the board JSON.
  /// MUST stay 1-to-1 with `TaskOutcomeKey` in `server/src/protocol.ts`.
  String get key => name;

  /// Reverse of [key]. Returns null for unknown strings so old saves with
  /// no outcome field stay loadable.
  static TaskOutcome? fromKey(String? raw) {
    if (raw == null) return null;
    for (final v in values) {
      if (v.name == raw) return v;
    }
    return null;
  }
}

/// `0.4 - 0.03 * precisionSkill`, clamped to `[0, 0.4]`.
/// * Precision 0 → 40% bug chance.
/// * Precision 7 → 19%.
/// * Precision 14+ → 0%.
double bugChance({required int precisionSkill}) =>
    (0.4 - 0.03 * precisionSkill).clamp(0.0, 0.4);

/// `0.02 * creativitySkill`, clamped to `[0, 1]`.
/// Only relevant on divergent task types (gated in [rollOutcome]).
double critChance({required int creativitySkill}) =>
    (0.02 * creativitySkill).clamp(0.0, 1.0);

/// `0.85 + 0.01 * reliabilitySkill`, clamped to `[0, 1]`.
/// * Reliability 0 → 15% chance of incomplete.
/// * Reliability 15+ → 0%.
double completionSuccessChance({required int reliabilitySkill}) =>
    (0.85 + 0.01 * reliabilitySkill).clamp(0.0, 1.0);

/// Task-type identifiers considered divergent (where creativity matters).
const divergentTaskTypes = <String>{
  'architecture',
  'product-spec',
  'ui-design',
};

/// Task types that require a physical workstation (computer). Agents without
/// an assigned desk suffer an incomplete-chance penalty on these.
const workstationTaskTypes = <String>{
  'coding',
  'testing',
  'debugging',
};

/// Penalty applied to [completionSuccessChance] when an agent has no desk
/// and is working on a [workstationTaskTypes] task. −0.25 success rate.
const double kUnassignedIncompletePenalty = 0.25;

/// Hard cap on the aggregate specialization crit bonus applied inside
/// [rollOutcome]. Mirrors `kMaxSpecializationCritBonus` in game_economy.dart;
/// duplicated here because [rollOutcome] is intentionally dependency-free
/// (no model imports). Keep the two values in sync.
const double kMaxSpecializationCritBonusInRoll = 0.30;

/// Hard cap on the lesson-driven success bonus applied inside [rollOutcome].
/// Mirrors `kMaxLessonSuccessBonus` in game_economy.dart. Keep in sync.
const double kMaxLessonSuccessBonusInRoll = 0.10;

/// Hard cap on the project memory depth bonus applied inside [rollOutcome].
/// Mirrors `kMaxProjectMemoryBonus` in game_economy.dart. Keep in sync.
const double kMaxProjectMemoryBonusInRoll = 0.15;


/// `kLessonSuccessBonusPerLesson * lessonCount`, clamped to
/// `[0, kMaxLessonSuccessBonusInRoll]`.
/// * 0 lessons → no bonus.
/// * 20 lessons → +10% success (= −10% incomplete rate).
double lessonSuccessBonus({required int lessonCount}) =>
    (0.005 * lessonCount).clamp(0.0, kMaxLessonSuccessBonusInRoll);

/// `0.01 * (totalTasksCompleted / 5)`, clamped to
/// `[0, kMaxProjectMemoryBonusInRoll]`.
/// * 0 tasks → no bonus.
/// * 75+ tasks → +15% crit on architecture rolls.
/// Applied only to `architecture` divergent tasks. Represents accumulated
/// project knowledge translating to better architectural insights.
double projectMemoryDepthBonus({required int totalTasksCompleted}) =>
    (0.01 * (totalTasksCompleted / 5)).clamp(0.0, kMaxProjectMemoryBonusInRoll);


/// Rolls the final outcome for a task at the `testing → done` transition.
///
/// Order of checks:
/// 1. Reliability gate (incomplete if we roll above the success threshold).
///    Agents without a workstation on [workstationTaskTypes] suffer a −0.25
///    penalty to their success chance.
///    [lessonBonus] (from accumulated lessons) adds to the success chance,
///    partially offsetting the base incomplete rate.
/// 2. Bug check (Precision).
/// 3. Crit check (Creativity + specialization, divergent tasks only).
///    [projectMemoryBonus] is added for `architecture` tasks specifically.
/// 4. Otherwise `clean`.
///
/// [specializationCritBonus] is the additive crit bonus from this agent's
/// unlocked specializations matching the current `taskType`. Caller computes
/// it; this function just clamps it defensively and folds it into the
/// crit chance.
///
/// [lessonBonus] is the additive success-chance bonus from accumulated lessons
/// (see [lessonSuccessBonus]). Caller computes it; clamped defensively here.
///
/// [projectMemoryBonus] is the additive crit bonus from accumulated project
/// knowledge (tasks completed). Applied only to `architecture` divergent tasks
/// (see [projectMemoryDepthBonus]). Caller computes it; clamped here.
///

/// [rng] is injectable for deterministic tests.
TaskOutcome rollOutcome({
  required Random rng,
  required int precisionSkill,
  required int creativitySkill,
  required int reliabilitySkill,
  bool isDivergentTask = false,
  bool isUnassigned = false,
  String taskType = '',
  double specializationCritBonus = 0.0,
  double lessonBonus = 0.0,
  double projectMemoryBonus = 0.0,
}) {
  final baseSuccess = completionSuccessChance(reliabilitySkill: reliabilitySkill);
  final needsDesk = isUnassigned && workstationTaskTypes.contains(taskType);
  final clampedLesson = lessonBonus.clamp(0.0, kMaxLessonSuccessBonusInRoll);
  final adjustedSuccess = needsDesk
      ? (baseSuccess + clampedLesson - kUnassignedIncompletePenalty).clamp(0.0, 1.0)
      : (baseSuccess + clampedLesson).clamp(0.0, 1.0);

  if (rng.nextDouble() > adjustedSuccess) {
    return TaskOutcome.incomplete;
  }
  if (rng.nextDouble() < bugChance(precisionSkill: precisionSkill)) {
    return TaskOutcome.bug;
  }
  if (isDivergentTask) {
    final base = critChance(creativitySkill: creativitySkill);
    final specBonus =
        specializationCritBonus.clamp(0.0, kMaxSpecializationCritBonusInRoll);
    final memoryBonus = taskType == 'architecture'
        ? projectMemoryBonus.clamp(0.0, kMaxProjectMemoryBonusInRoll)
        : 0.0;
    if (rng.nextDouble() < (base + specBonus + memoryBonus).clamp(0.0, 1.0)) {
      return TaskOutcome.crit;
    }
  }
  return TaskOutcome.clean;
}

File: server/src/task_outcome.ts
/**
 * Pure functions that roll a task's final outcome based on agent skills.
 *
 * Server-side mirror of `lib/services/task_outcome.dart`. Kept 1-to-1 with
 * the Dart implementation (same formulas, same clamps, same constants) so
 * the testing → ? roll can move from the client simulator to the server
 * (C.2 "Board transitions — server as single writer") without changing
 * the empirically-tuned outcome distribution.
 *
 * If you touch any constant or formula here, update the Dart file in the
 * same commit and re-run both test suites — drift between the two will
 * cause silent gameplay regressions.
 */

export type TaskOutcome = "clean" | "crit" | "bug" | "incomplete";

/** RNG that yields [0, 1). Injected so tests can pin a seeded sequence. */
export type Rng = () => number;

// ─── Skill-driven chance helpers ─────────────────────────────────────────────

/** `0.4 - 0.03 * precisionSkill`, clamped to `[0, 0.4]`. */
export function bugChance(precisionSkill: number): number {
  return clamp(0.4 - 0.03 * precisionSkill, 0.0, 0.4);
}

/** `0.02 * creativitySkill`, clamped to `[0, 1]`. Divergent tasks only. */
export function critChance(creativitySkill: number): number {
  return clamp(0.02 * creativitySkill, 0.0, 1.0);
}

/** `0.85 + 0.01 * reliabilitySkill`, clamped to `[0, 1]`. */
export function completionSuccessChance(reliabilitySkill: number): number {
  return clamp(0.85 + 0.01 * reliabilitySkill, 0.0, 1.0);
}

// ─── Task-type sets ──────────────────────────────────────────────────────────

export const divergentTaskTypes: ReadonlySet<string> = new Set([
  "architecture",
  "product-spec",
  "ui-design",
]);

export const workstationTaskTypes: ReadonlySet<string> = new Set([
  "coding",
  "testing",
  "debugging",
]);

// ─── Bonus caps (mirror game_economy.dart) ───────────────────────────────────

export const kUnassignedIncompletePenalty = 0.25;
export const kMaxSpecializationCritBonusInRoll = 0.30;
export const kMaxLessonSuccessBonusInRoll = 0.10;
export const kMaxProjectMemoryBonusInRoll = 0.15;

/** `0.005 * lessonCount`, clamped to `[0, kMaxLessonSuccessBonusInRoll]`. */
export function lessonSuccessBonus(lessonCount: number): number {
  return clamp(0.005 * lessonCount, 0.0, kMaxLessonSuccessBonusInRoll);
}

/** `0.01 * (totalTasksCompleted / 5)`, clamped. Architecture tasks only. */
export function projectMemoryDepthBonus(totalTasksCompleted: number): number {
  return clamp(0.01 * (totalTasksCompleted / 5), 0.0, kMaxProjectMemoryBonusInRoll);
}

// ─── Roll ────────────────────────────────────────────────────────────────────

export interface RollInput {
  rng: Rng;
  precisionSkill: number;
  creativitySkill: number;
  reliabilitySkill: number;
  isDivergentTask?: boolean;
  isUnassigned?: boolean;
  taskType?: string;
  specializationCritBonus?: number;
  lessonBonus?: number;
  projectMemoryBonus?: number;
}

/**
 * Order of checks (mirror Dart):
 *   1. Reliability gate (incomplete if we roll above the success threshold).
 *      Agents without a workstation on workstation tasks suffer a −0.25 penalty.
 *      `lessonBonus` adds to success, partially offsetting incomplete rate.
 *   2. Bug check (precision).
 *   3. Crit check (creativity + specialization, divergent tasks only).
 *      `projectMemoryBonus` is added for architecture tasks specifically.
 *   4. Otherwise `clean`.
 */
export function rollOutcome(input: RollInput): TaskOutcome {
  const {
    rng,
    precisionSkill,
    creativitySkill,
    reliabilitySkill,
    isDivergentTask = false,
    isUnassigned = false,
    taskType = "",
    specializationCritBonus = 0.0,
    lessonBonus = 0.0,
    projectMemoryBonus = 0.0,
  } = input;

  const baseSuccess = completionSuccessChance(reliabilitySkill);
  const needsDesk = isUnassigned && workstationTaskTypes.has(taskType);
  const clampedLesson = clamp(lessonBonus, 0.0, kMaxLessonSuccessBonusInRoll);
  const adjustedSuccess = needsDesk
    ? clamp(baseSuccess + clampedLesson - kUnassignedIncompletePenalty, 0.0, 1.0)
    : clamp(baseSuccess + clampedLesson, 0.0, 1.0);

  if (rng() > adjustedSuccess) {
    return "incomplete";
  }
  if (rng() < bugChance(precisionSkill)) {
    return "bug";
  }
  if (isDivergentTask) {
    const base = critChance(creativitySkill);
    const specBonus = clamp(
      specializationCritBonus,
      0.0,
      kMaxSpecializationCritBonusInRoll,
    );
    const memoryBonus = taskType === "architecture"
      ? clamp(projectMemoryBonus, 0.0, kMaxProjectMemoryBonusInRoll)
      : 0.0;
    if (rng() < clamp(base + specBonus + memoryBonus, 0.0, 1.0)) {
      return "crit";
    }
  }
  return "clean";
}

// ─── Helpers ─────────────────────────────────────────────────────────────────

function clamp(v: number, lo: number, hi: number): number {
  if (v < lo) return lo;
  if (v > hi) return hi;
  return v;
}

// ─── Seedable RNG for tests ──────────────────────────────────────────────────

/**
 * Mulberry32 — small, fast, well-distributed PRNG. Used in tests to pin a
 * deterministic outcome sequence; in production the runner injects
 * `Math.random` directly.
 *
 * Bit-identical between Node and any other JS runtime: only uses uint32 ops.
 */
export function mulberry32(seed: number): Rng {
  let s = seed >>> 0;
  return () => {
    s = (s + 0x6d2b79f5) >>> 0;
    let t = s;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}
